src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
```

Comment:
I know we discussed this, but I still find it unintuitive that a `SrcLoc` is needed for `case` and `type-case`, but not for anything else. We should document the reasoning somewhere. Let's see, that reasoning is...

That `SrcLoc` is used to construct the error message in the case in which no cases match. By comparison, other errors such as looking up a `Var` don't need a `SrcLoc` because scope errors are caught during expansion time, before evaluation begins. Are case and type-case really the only ways in which evaluation can fail at runtime, assuming expansion already caught scope errors and type errors?

Reply:
Actually, the current datatype exhaustively lists all the errors which can happen at evaluation-time, and #237 already classifies which ones should already be caught in an earlier phase and which ones really can happen at evaluation-time. There is `EvalErrorCase`, and `EvalErrorUser`. `EvalErrorUser` does not need a separate `SrcLoc` because the error message is a `Syntax`, which is already associated with a `SrcLoc`.

Since we're separating between `Case` and `TypeCase` everywhere else in the codebase, I wonder why there isn't a separate `EvalErrorTypeCase` here? :thinking:

Anyway, that's not something which changed in the current MR, moving right along...